db/version_set.cc

```diff
@@ -4881,59 +4881,112 @@ Status VersionSet::WriteSnapshot(log::Writer* log) {
 uint64_t VersionSet::ApproximateSize(Version* v, const Slice& start,
                                      const Slice& end, int start_level,
                                      int end_level, TableReaderCaller caller) {
+  const auto& icmp = v->cfd_->internal_comparator();
+
   // pre-condition
-  assert(v->cfd_->internal_comparator().Compare(start, end) <= 0);
+  assert(icmp.Compare(start, end) <= 0);
 
   uint64_t size = 0;
   const auto* vstorage = v->storage_info();
-  end_level = end_level == -1
-                  ? vstorage->num_non_empty_levels()
-                  : std::min(end_level, vstorage->num_non_empty_levels());
+  const int num_non_empty_levels = vstorage->num_non_empty_levels();
+  end_level = (end_level == -1) ? num_non_empty_levels
+                                : std::min(end_level, num_non_empty_levels);
 
   assert(start_level <= end_level);
 
-  for (int level = start_level; level < end_level; level++) {
+  std::vector<FdWithKeyRange*> first_files;
+  first_files.reserve(end_level - start_level);
+
+  std::vector<FdWithKeyRange*> last_files;
+  last_files.reserve(end_level - start_level);
+
+  // scan all the levels in reverse order
+  for (int level = end_level - 1; level >= start_level; --level) {
     const LevelFilesBrief& files_brief = vstorage->LevelFilesBrief(level);
-    if (!files_brief.num_files) {
+    if (files_brief.num_files == 0) {
       // empty level, skip exploration
       continue;
     }
 
-    if (!level) {
-      // level 0 data is sorted order, handle the use case explicitly
+    if (level == 0) {
```

Comment:
I think we can just put all L0 files into `first_files` list.

Reply:
It's a bit tricky because of lack of ordering.
most of the ApproximateSize calls inside ApproximateSizeLevel0 are going to be cheap (returning 0 or file_size). I don't feel that the code complication is worth it, and would rather leave it as is, unless u insist on another approach.

Reply:
ApproximateSizeLevel0() does exactly the same as what we do in `first_files`. What's the difference?

Reply:
will explain below

Reply:
hmm... can't add comments below. explaining here.
they do ALMOST the same.
the difference is whether we use the heuristic or not.
if i move ApproximateSizeLevel0 to first_files handling, it could happen that if total_full_size is big enough - we will approximate ALL L0 files as half of their size (not only the boundary files).

to me - it feels like not worth the trouble to do this.